forge/src/main/java/me/lucko/luckperms/forge/mixin/core/commands/CommandsMixin.java

```diff
@@ -0,0 +1,91 @@
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge.mixin.core.commands;
+
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.tree.CommandNode;
+import com.mojang.brigadier.tree.RootCommandNode;
+import me.lucko.luckperms.forge.event.SuggestCommandsEvent;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.Commands;
+import net.minecraft.commands.SharedSuggestionProvider;
+import net.minecraftforge.common.MinecraftForge;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Redirect;
+
+import java.util.Map;
+
+/**
+ * Mixin into {@link Commands} for posting {@link SuggestCommandsEvent}
+ */
+@Mixin(value = Commands.class)
+public abstract class CommandsMixin {
+
+    @Shadow
+    protected abstract void fillUsableCommands(CommandNode<CommandSourceStack> p_82113_, CommandNode<SharedSuggestionProvider> p_82114_, CommandSourceStack p_82115_, Map<CommandNode<CommandSourceStack>, CommandNode<SharedSuggestionProvider>> p_82116_);
+
+    @Redirect(
+            method = "sendCommands",
+            at = @At(
+                    value = "INVOKE",
+                    target = "Lnet/minecraft/commands/Commands;fillUsableCommands(Lcom/mojang/brigadier/tree/CommandNode;Lcom/mojang/brigadier/tree/CommandNode;Lnet/minecraft/commands/CommandSourceStack;Ljava/util/Map;)V"
+            )
+    )
+    private void onFillUsableCommands(Commands commands, CommandNode<CommandSourceStack> commandNode, CommandNode<SharedSuggestionProvider> suggestionNode, CommandSourceStack source, Map<CommandNode<CommandSourceStack>, CommandNode<SharedSuggestionProvider>> map) {
+        SuggestCommandsEvent event = new SuggestCommandsEvent(source, (RootCommandNode<CommandSourceStack>) commandNode);
+        MinecraftForge.EVENT_BUS.post(event);
+
+        // This map will be populated with the original root node, so we must clear it
+        map.clear();
+
+        // Insert the root node as it may have been replaced during the event
+        map.put(event.getNode(), suggestionNode);
+
+        fillUsableCommands(event.getNode(), suggestionNode, source.withPermission(4), map);
+    }
+
+    /**
+     * Forge will only re-throw the Exception passed in the {@link net.minecraftforge.event.CommandEvent CommandEvent} if it's an unchecked Exception,
+     * this behaviour prevents a {@link CommandSyntaxException CommandSyntaxException} from being propagated through the event.
+     *
+     * TODO Fix upstream.
```

Comment:
This Mixin can be removed when this is merged: https://github.com/MinecraftForge/MinecraftForge/pull/8590